core/src/main/scala/kafka/server/metadata/BrokerMetadataEventProcessor.scala

```diff
@@ -0,0 +1,156 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package kafka.server.metadata
+
+import kafka.coordinator.group.GroupCoordinator
+import kafka.coordinator.transaction.TransactionCoordinator
+import kafka.server.{KafkaConfig, MetadataCache, QuotaFactory, ReplicaManager}
+import kafka.utils.Logging
+import org.apache.kafka.common.internals.FatalExitError
+import org.apache.kafka.common.protocol.ApiMessage
+
+/**
+ * Ability to process ApiMessage
+ */
+trait ApiMessageProcessor {
+  /**
+   * Process a list of messages
+   *
+   * @param apiMessages the messages to process
+   */
+  def process(apiMessages: List[ApiMessage]): Unit
+
+  /**
+   * Process a register-local-broker event from the broker's heartbeat
+   *
+   * @param registerLocalBrokerEvent the event to process
+   */
+  def process(registerLocalBrokerEvent: OutOfBandRegisterLocalBrokerEvent): Unit
+
+  /**
+   * Process a fence-local-broker event from the broker's heartbeat
+   *
+   * @param fencerLocalBrokerEvent the event to process
+   */
+  def process(fencerLocalBrokerEvent: OutOfBandFenceLocalBrokerEvent): Unit
+}
+
+object BrokerMetadataEventProcessor {
+  // visible for testing
+  @volatile private[metadata] var currentMetadataOffset: Long = -1
+
```

Comment:
Also, this can't be a static variable.  A lot of JUnit tests feature multiple brokers.